### PR TITLE
Private Collection and Sum Transform in Beam's public API.

### PR DESCRIPTION
**Pull request by preethiraghavan1**
## Description
Implemented PrivateCollection, MakePrivate transform and Sum transform for Beam public API.

## How has this been tested?
- Unit tests

## Checklist
- [ x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/private_beam.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+class MakePrivate(PrivateTransform):
+    """
+    Transform class for creating a PrivateCollection.
+    """
+
+    def __init__(self,
+                 budget_accountant: budget_accounting.BudgetAccountant,
+                 privacy_id_extractor: Callable,
+                 label: Optional[str] = None):
+        super(MakePrivate, self).__init__(label)
```
Comment (dvadym) [inline pipeline_dp/private_beam.py:86]: Nit: `super().__init__`  and in other places

In Python 3 no need to specify base class

File: pipeline_dp/private_beam.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            low=self._sum_params.low,
+            high=self._sum_params.high,
+            public_partitions=self._sum_params.public_partitions)
+
+        data_extractors = pipeline_dp.DataExtractors(
+            partition_extractor=self._sum_params.partition_extractor,
+            privacy_id_extractor=self.privacy_id_extractor,
+            value_extractor=self._sum_params.value_extractor)
+
+        dp_result = dp_engine.aggregate(pcol, params, data_extractors)
```
Comment (dvadym) [inline pipeline_dp/private_beam.py:127]: NIt: `return dp_engine.agg...`

File: tests/private_beam_test.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        return pcol | "Identity transform" >> beam.Map(lambda x: x)
+
+
+class PrivateBeamTest(unittest.TestCase):
+
+    @staticmethod
+    def privacy_id_extractor(x):
+        return "pid:" + x
+
+    def test_MakePrivate_transform_succeeds(self):
```
Comment (dvadym) [inline tests/private_beam_test.py:24]: Wow, thanks for thorough testing!

File: tests/private_beam_test.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                    budget_accountant=budget_accountant,
+                    privacy_id_extractor=PrivateBeamTest.privacy_id_extractor))
+
+            # Act
+            transformed = private_collection | SimplePrivateTransform()
+
+            # Assert
+            self.assertIsInstance(transformed, PCollection)
+
+    @patch('pipeline_dp.dp_engine.DPEngine.aggregate')
```
Comment (dvadym) [inline tests/private_beam_test.py:112]: Nice, that's the right way to test, to ensure that DPEngine is called in correct way!

**Review by dvadym — COMMENTED** [review]
Thanks for implementing PrivatePCollection it looks great!

I left some comments

File: pipeline_dp/private_beam.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+from abc import abstractmethod
+from typing import Callable, Optional
+from apache_beam.pvalue import PCollection
+
+import pipeline_dp
+from pipeline_dp import aggregate_params, budget_accounting
+
+
+class PrivateTransform(ptransform.PTransform):
+    """
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/private_beam.py]: Done!

File: pipeline_dp/private_beam.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    Abstract class for PrivateTransforms.
+    """
+
+    def __init__(self, label: Optional[str] = None):
+        super(PrivateTransform, self).__init__(label)
+        self.return_private = False
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/private_beam.py]: Done!

File: pipeline_dp/private_beam.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        self.budget_accountant = None
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/private_beam.py]: Done!

File: pipeline_dp/private_beam.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+class PrivateCollection:
+    """
+    Private counterpart for PCollection.
+    """
+
+    def __init__(self, pcol: PCollection,
+                 budget_accountant: budget_accounting.BudgetAccountant,
+                 privacy_id_extractor: Callable):
+        self.pcol = pcol
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/private_beam.py]: Done!

File: pipeline_dp/private_beam.py (hunk 1 above)
Comment (preethiraghavan1) [reply pipeline_dp/private_beam.py]: Done!

File: pipeline_dp/private_beam.py (hunk 2 above)
Comment (preethiraghavan1) [reply pipeline_dp/private_beam.py]: Done!

File: tests/private_beam_test.py (hunk 4 above)
Comment (preethiraghavan1) [reply tests/private_beam_test.py:111]: Thanks!

File: tests/private_beam_test.py (hunk 3 above)
Comment (preethiraghavan1) [reply tests/private_beam_test.py:24]: :) Thanks!

File: pipeline_dp/private_beam.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        return PrivateCollection(pcol, self._budget_accountant,
+                                 self._privacy_id_extractor)
+
+
+class Sum(PrivateTransform):
+    """Transform class for performing DP Sum on PrivateCollection."""
+
+    def __init__(self,
+                 sum_params: aggregate_params.SumParams,
+                 return_private: bool,
```
Comment (dvadym) [inline pipeline_dp/private_beam.py:78]: `Sum` returns anonymized result, so no need to specify return_anonymized as argument, it's always `true`.

The next task will be to implement `Map` which maps `PrivatePCollection` to `PrivatePCollection`, i.e. `return_anonymized` will be `false`.

File: pipeline_dp/private_beam.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        self._budget_accountant = budget_accountant
+        self._privacy_id_extractor = privacy_id_extractor
+
+    @abstractmethod
+    def expand(self, pcol: pvalue.PCollection) -> pvalue.PCollection:
+        pass
+
+
+class PrivateCollection:
+    """Private counterpart for PCollection."""
```
Comment (dvadym) [inline pipeline_dp/private_beam.py:32]: Please add:

PrivateCollection guarantees that only anonymized data within the specified privacy budget can be extracted from it through PrivatePTransforms.

**Review by dvadym — COMMENTED** [review]
A few small comments

File: pipeline_dp/private_beam.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+import pipeline_dp
+from pipeline_dp import aggregate_params, budget_accounting
+
+
+class PrivateTransform(ptransform.PTransform):
+    """Abstract class for PrivateTransforms."""
+
+    def __init__(self, return_private: bool, label: Optional[str] = None):
+        super().__init__(label)
+        self._return_private = return_private
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/private_beam.py:15]: So, if return_anonymized = True, then return PCollection else PrivatePCollection

It is opposite of return_private where 
if return_private= True, then return PrivatePCollection else PCollection.

In that case, Sum should return PCollection and have return_anonymized= False.
Did I understand correctly?

**Pull request merged**